metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/events/TaxonomyEventHandler.java

```diff
@@ -77,6 +77,9 @@ public Taxonomy importTaxonomyTree(Taxonomy taxonomy) throws Exception {
                 long taxIdLong = Long.parseLong(taxId);
                 Taxonomy taxonomyParent = importTaxonomyTree(new Taxonomy(taxIdLong));
                 taxonomy.setTaxonomyForRank(taxonomyParent, rankEnum.toString());
+            } else if (taxonomy.getRank().equals(RANK.SPECIES.name) && taxonomy.getTaxonomySpecies() == null) {
+                // self-reference at species level
```

Comment:
About the data integrity violation exception: I have a feeling that you have to save an entity before you can self-reference it. So I suggest you rewrite it in such a way that the entity is first saved (with an empty `taxonomySpecies` field), and _then_ the `setTaxonomyForRank` is called.

Reply:
(As discussed in Slack, this does not solve the problem of the data integrity violation.)

OK, so I've tested the current code (as it is in this PR right now), and I see that the actual self-referencing actually does work. For example, if you import a study ERP016904, you get the species taxonomy which correctly refers back to itself in `taxonomySpecies`. You can view it in the metadata web service with no problem. So, first of all, we don't need to save the taxonomy prior to self-referencing it; and secondly, the problem appears to be not in the self-referencing part at all.

The error thrown in the tests is for the study ERP000054. It doesn't matter if you import it alone or with other studies, it always throws an error:

```
SEVERE: Encountered Exception for study accession ERP000054
Nov 06, 2019 9:57:00 AM uk.ac.ebi.ampt2d.metadata.importer.MetadataImporterMainApplication lambda$run$1
SEVERE: An analysis must be mapped against one or more reference sequences
2019-11-06 09:57:52.279  WARN   --- [           main] o.h.engine.jdbc.spi.SqlExceptionHelper   : SQL Error: 90053, SQLState: 90053
2019-11-06 09:57:52.281 ERROR   --- [           main] o.h.engine.jdbc.spi.SqlExceptionHelper   : Scalar subquery contains more than one row; SQL statement:
select study0_.id as id1_9_0_, study0_.last_modified_date as last_mod2_9_0_, ...
[Long SQL statement redacted for brevity]
```

It looks like the code which checks the validity of an analysis is doing something wrong (in a way not compatible with self referencing). Since this part was recently changed, it seems likely. So I suggest you look into that part

Reply:
Oh, wait a minute, the issue with ERP000054 is that its analysis https://www.ebi.ac.uk/ena/data/view/ERZ000001&display=xml indeed does not have a reference sequence, so this is *why* we're using at as an example of a study with an invalid analysis. So the problem is not that, I must have misread the logs